### DISCUSSION TITLE
Do a list of "agents" features

### DISCUSSION THREAD

**Discussion by SergeStinckwich** [cormas/cormas#774] [General]
@ChLePage Actually the existence of such a large number of classes is due to fact that it was decided to create as many different classes as there interesting combinations of agent “features”. We have two main features for an agent in Cormas: communication through mailboxes and location on an elementary spatial entity (patch). A specific agent needing to be equipped with communication and location features could not inherit from both AgentComm and AgentLocation… Therefore it was decided to create AgentCommLocation… As long as a Cormas user can continue to select the needed features by checking boxes from the entity-creation interface, wa can discuss about a more elegant-efficient way of handling this I suppose …

**Comment by SergeStinckwich** [comment]
In fact traits in Pharo is a nice way to solve this problem. For example, look at TSortable, that represent methods that are related to element sorting.A Trait is a set of method, i.e they only contain behavior. A class can combine many traits in order to implement multiple inheritance.

In your case, we can define ``TLocation`` amd ``TComm`` traits and define a CMAgentCommLocation class that inherit from both traits. The main interest is that you don't have to duplicate code. When you modify a traits, all the classes using it are modified immediately. The hierarchy is less complex.

You have also Talents in Pharo that are some kind of traits that you could attach to objects (not classes), not sure we want to have an agent that dynamically acquire a communication talent at one point. See more details here: https://github.com/tesonep/pharo-talents

**Comment by SergeStinckwich** [comment]
Can we do a list of features that we could combine then?

**Comment by SergeStinckwich** [comment]
- Comm
- Location
- Group (but not used apparently)

**Comment by nbecu** [comment]
Comm Location and Group are the three features for agents. It's also possible to have a "simple agent" that doesn't have any of these features. 

Such a "simple agent" is very close too what's named PassiveEntity (or PassiveObject)

**Comment by nbecu** [comment]
Continuing on this topic PassiveObject have one  optional feature, which is Location

Note that there is a slight difference between a Agent with the feature Location and a PassiveObject with the feature Location.   A Agent with Location, is considered to be able to move autonomously, so he has the message moveTo:
A PassiveObject with Location feature, is considered not to be autonomous, so he has the message isMovedTo:

(In recent years users tend to use a hack to allow polymorphism between Agent and PassiveObject, by introducing a moveTo: message in the structure of PassiveObject-Location Class)

**Comment by nbecu** [comment]
Finally, Spatial Entity have several types and/or features as well. It can be a Cell or a ComposedSpatialEntity.  A Cell can be added the feature CellularAutomataCell, which will give it the ability to respond to updateState (I'm not sure of the exact message name). 
A ComposedSpatialEntity is an abstract class. It has two subclasses that can be instantiate. Aggregate and NonConnexSpatialEntity (I'm not sure of the name).  Those two classes are not really features, but more types of entities. 


Hope these comments help. 
Cheers
